src/java.base/share/classes/java/lang/foreign/Linker.java

```diff
@@ -302,70 +573,71 @@ static Option firstVariadicArg(int index) {
+         * This parameter, called the 'capture state segment', represents the native segment into which
+         * the captured state is written.
          * <p>
-         * The native segment should have the layout {@linkplain CaptureCallState#layout associated}
-         * with the particular {@code CaptureCallState} instance used to link the downcall handle.
+         * The capture state segment should have the layout returned by {@linkplain #captureStateLayout}.
+         * This layout is a struct layout which has a named field for each captured value.
          * <p>
-         * Captured state can be retrieved from this native segment by constructing var handles
-         * from the {@linkplain #layout layout} associated with the {@code CaptureCallState} instance.
+         * Captured state can be retrieved from the capture state segment by constructing var handles
+         * from the {@linkplain #captureStateLayout capture state layout}.
          * <p>
          * The following example demonstrates the use of this linker option:
          * {@snippet lang = "java":
          * MemorySegment targetAddress = ...
-         * CaptureCallState ccs = Linker.Option.captureCallState("errno");
+         * Linker.Option ccs = Linker.Option.captureCallState("errno");
          * MethodHandle handle = Linker.nativeLinker().downcallHandle(targetAddress, FunctionDescriptor.ofVoid(), ccs);
          *
-         * VarHandle errnoHandle = ccs.layout().varHandle(PathElement.groupElement("errno"));
-         * try (Arena arena = Arena.openConfined()) {
-         *     MemorySegment capturedState = arena.allocate(ccs.layout());
+         * StructLayout capturedStateLayout = Linker.Option.capturedStateLayout();
+         * VarHandle errnoHandle = capturedStateLayout.varHandle(PathElement.groupElement("errno"));
+         * try (Arena arena = Arena.ofConfined()) {
+         *     MemorySegment capturedState = arena.allocate(capturedStateLayout);
          *     handle.invoke(capturedState);
          *     int errno = errnoHandle.get(capturedState);
          *     // use errno
          * }
          * }
+         *
+         * @param capturedState the names of the values to save.
+         * @see #captureStateLayout()
+         */
+        static Option captureCallState(String... capturedState) {
+            Set<CapturableState> set = Stream.of(Objects.requireNonNull(capturedState))
+                    .map(Objects::requireNonNull)
+                    .map(CapturableState::forName)
+                    .collect(Collectors.toSet());
+            return new LinkerOptions.CaptureCallState(set);
+        }
+
+         /**
+         * {@return A struct layout that represents the layout of the capture state segment that is passed
+         *          to a downcall handle linked with {@link #captureCallState(String...)}}
+         *
+         * @see #captureCallState(String...)
```

Comment:
How does a caller know what the structure may contain? Should we document the platform specific structures?

Reply:
I've added an example of how to print the platform-dependent structure. Should we document anyhow?

Reply:
I'm not sure about this. Honestly, the example probably doesn't help much if somebody didn't get the idea of what the layout might be. I think it might be better to say something like, `on Windows/x64 the returned layout might be as follows...` and then you write the code to create the layout instance corresponding to the returned layout. But we have to be careful to make the text non-normative (as the set of values might change).

Reply:
What about adding something like this?

```
Here is a list of valid names for some platforms:
    Linux:
        "errno"
    MacOS:
       "errno"
    Windows:
        "GetLastError", "WSAGetLastError" and "errno"
```

Reply:
Back to @PaulSandoz  question - "how does the caller know what the structure contains?". The caller typically doesn't care too much about what the returned struct is. But it might care about which "values" can be captured. That said, the set of such interesting values, is not too surprising. As demonstrated in the example in the `Linker.capturedCallState` method, once the client knows the name (and "errno" is likely to be 90% case), everything else follows from there - as the layout can be used to obtain var handles for the required values.

But, perhaps, now that I write this, I realize that what @PaulSandoz might _really_ be asking is: how do I know that e.g. the returned struct will not contain e.g. nested structs, sequences, or other non-sense. So we might specify (in a normative way) that the returned layout is a struct layout, whose member layouts are one or more value layouts (possibly with some added padding layouts). The names of the value layouts reflect the names of the values that can be captured.

Reply:
This is much clearer. Are the names of the struct members guaranteed to be symbols that can be found up via the linker's default lookup? i.e. a small set of important global variables.

Reply:
While they might be defined as global variable, I wouldn't read too much into this. Even `errno` is not really a global variable, but some thread local value that the native compiler knows about (e.g. setting the global variable doesn't really do much). Quoting from Linux man page:

```
errno  is  defined  by  the ISO C standard to be a modifiable lvalue of
       type int, and must not be explicitly declared; errno may  be  a  macro.
       errno  is  thread-local;  setting  it in one thread does not affect its
       value in any other thread.
```

This seems to imply that errno might even be something else. Given how the story goes for errno, I don't think we'd want to restrict the spec to state that these names should be discoverable via the lookup.